engine/gamesys/src/gamesys/components/comp_particlefx.cpp

```diff
@@ -408,24 +408,25 @@ namespace dmGameSystem
     static void RenderListDispatch(dmRender::RenderListDispatchParams const &params)
     {
         ParticleFXWorld* pfx_world = (ParticleFXWorld*)params.m_UserData;
-
-        if (params.m_Operation == dmRender::RENDER_LIST_OPERATION_BEGIN)
-        {
-            pfx_world->m_VertexBufferData.SetSize(0);
-            pfx_world->m_RenderObjects.SetSize(0);
-        }
-        else if (params.m_Operation == dmRender::RENDER_LIST_OPERATION_BATCH)
+        switch(params.m_Operation)
         {
-            RenderBatch(pfx_world, params.m_Context, params.m_Buf, params.m_Begin, params.m_End);
-        }
-        else if (params.m_Operation == dmRender::RENDER_LIST_OPERATION_END)
-        {
-            dmRender::SetBufferData(params.m_Context, pfx_world->m_VertexBuffer, pfx_world->m_VertexBufferData.Size(), pfx_world->m_VertexBufferData.Begin(), dmGraphics::BUFFER_USAGE_STREAM_DRAW);
-
-            DM_PROPERTY_ADD_U32(rmtp_ParticleVertexCount, pfx_world->m_VerticesWritten);
-            DM_PROPERTY_ADD_U32(rmtp_ParticleVertexSize, pfx_world->m_VertexBufferData.Size());
-
-            pfx_world->m_DispatchCount++;
+            case dmRender::RENDER_LIST_OPERATION_BEGIN:
+                pfx_world->m_VertexBufferData.SetSize(0);
+                pfx_world->m_RenderObjects.SetSize(0);
+                break;
+            case dmRender::RENDER_LIST_OPERATION_BATCH:
+                RenderBatch(pfx_world, params.m_Context, params.m_Buf, params.m_Begin, params.m_End);
+                break;
+            case dmRender::RENDER_LIST_OPERATION_END:
+                if (pfx_world->m_VertexBufferData.Size() > 0)
+                {
+                    dmRender::SetBufferData(params.m_Context, pfx_world->m_VertexBuffer, pfx_world->m_VertexBufferData.Size(), pfx_world->m_VertexBufferData.Begin(), dmGraphics::BUFFER_USAGE_STREAM_DRAW);
+                    DM_PROPERTY_ADD_U32(rmtp_ParticleVertexCount, pfx_world->m_VerticesWritten);
+                    DM_PROPERTY_ADD_U32(rmtp_ParticleVertexSize, pfx_world->m_VertexBufferData.Size());
+                    pfx_world->m_DispatchCount++;
```

Comment:
(Not saying this is part of this PR)

Q: Is this working the way we want it to? (perhaps we've postponed it to a MVP2 or something?)
(Assuming we want the same logic in all components, so think "model component" as a current example, as it supports vertex formats.)

If I have 2 components with two materials, using the same predicate tag but different vertex formats.
During one `render.draw(pred)` I would get two calls to `RENDER_LIST_OPERATION_BATCH`, each requiring a different vertex buffer. Currently, this doesn't seem to match the code, which uses the "dispatch count", which essentially counts numbers of `render.draw()` calls?

To me it seems we need to manage buffers depending on `RENDER_LIST_OPERATION_BATCH` ?

Or is my logic way off here?

Reply:
In the case where you have two materials for one draw call, we can reuse the same vertex buffer regardless of how the vertex format looks like. The memory overwrite issue happens when you do multiple render.draw executions with the same component type:

```
render.draw(pred_1):
batch_begin (reset write pointers)
batch_1 (put vertices into vx buffer on cpu)
...
batch_x
batch_end (write vertices into vx buffer on GPU)

render.draw(pred_2):
batch_begin (again, reset write pointers)
batch_1 -> batch_x (same as before)
batch_end (write vertices on GPU with the SAME vx buffer handle as before)
```
The second batch_end call will start copying over the data to the memory mapped GPU buffer during the same frame, meaning we can get random results depending on when the data is read or written.

In the case of models, we do things a bit differently but it should essentially be the same thing (I'll have to double check why we don't advance the render buffers for models, might be something that's overlooked..)

Reply:
"we can reuse the same vertex buffer "

Really? Isn't the vertex buffer tied to a single vertex format? And, once the first batch is submitted, it is in flight and we don't want to alter it anymore?
I am probably missing something 🤔

Reply:
Ah, I'm talking about a different scenario:

```
render.draw(pred):
batch_begin (reset write pointers)
batch_1 (material 1, vertex format 1)
batch_2 (material 2, vertex format 2)
...
batch_x
batch_end
```

This to me seems like we don't handle well at the current moment?

Reply:
We do yes, we can reuse the same vertex buffer as long as we calculate all the offsets correct (they must start at an even vertex, which depends on the vertex declaration stride).